src/main/java/com/kodilla/ecommercee/controller/ProductController.java

```diff
@@ -1,70 +1,57 @@
-        }
-    };
+    @Autowired
+    ProductService productService;
+    @Autowired
+    ProductMapper productMapper;
 
     @GetMapping
     public ResponseEntity<List<ProductDto>> getProducts() {
-        return ResponseEntity.ok(testProductList);
+        List<Product> productList = productService.getAllProducts();
+        return ResponseEntity.ok(productMapper.mapToProductDtoList(productList));
     }
-
     @GetMapping("{productId}")
-    public ResponseEntity<ProductDto> getProduct(@PathVariable Long productId) {
-        for (ProductDto product : testProductList) {
-            if (product.getProductId().equals(productId)) {
-                return ResponseEntity.ok(product);
-            }
-        }
-        return ResponseEntity.notFound().build();
+    public ResponseEntity<ProductDto> getProduct(@PathVariable Long productId) throws ProductNotFoundException {
+        return ResponseEntity.ok(productMapper.mapToProductDto(productService.getProduct(productId)));
     }
-
     @DeleteMapping(value = "{productId}")
     public ResponseEntity<Void> deleteProduct(@PathVariable Long productId) {
-        for (ProductDto product : testProductList) {
-            if (product.getProductId().equals(productId)) {
-                testProductList.remove(product);
-                return ResponseEntity.ok().build();
-            }
-        }
-        return ResponseEntity.notFound().build();
+        productService.deleteById(productId);
+        return ResponseEntity.ok().build();
     }
-
     @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE)
     public ResponseEntity<Void> createProduct(@RequestBody ProductDto productDto) {
-        testProductList.add(productDto);
-
+        Product product = productMapper.mapToProduct(productDto);
+        productService.saveProduct(product);
         return ResponseEntity.ok().build();
     }
-
+    @PatchMapping(path = "{productId}", consumes = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<ProductDto> updateProduct(@PathVariable Long productId, @RequestBody ProductDto newProductDto) {
+        Product newProduct = productMapper.mapToProduct(newProductDto);
+        productService.saveProduct(newProduct);
+        return ResponseEntity.ok(productMapper.mapToProductDto(newProduct));
+    }
     @PutMapping(path = "{productId}", consumes = MediaType.APPLICATION_JSON_VALUE)
-    public ResponseEntity<ProductDto> updateProduct(@PathVariable Long productId, @RequestBody ProductDto productNew) {
-        for (ProductDto product : testProductList) {
-            if (product.getProductId().equals(productId)) {
-                testProductList.remove(product);
-                testProductList.add(productNew);
-                return ResponseEntity.ok(productNew);
-            }
-        }
-        return ResponseEntity.notFound().build();
+    public ResponseEntity<ProductDto> updateAndNewProduct(@PathVariable Long productId, @RequestBody ProductDto newProductDto) {
```

Comment:
Out of curiosity - what is this method used for?
Is it just to be available just in case?

Reply:
I think it could be useful, when we want to replace product by another (new version of product for example...). Maybe not so strong necessary, but I'd like to leave it for now...

Reply:
Sure... It was absolutely not a reservation on my part. I was just curious as to why. Now I completely agree and understand.